### PR TITLE
Add configuration forms for flood control of users and contact form.

### PR DESCRIPTION
**Pull request by herbdool**
Addresses https://github.com/backdrop/backdrop-issues/issues/3110

### CODE REVIEW

File: core/modules/user/user.admin.inc [hunk 1]
```diff
@@ -288,6 +288,94 @@ function user_admin_settings_submit($form, &$form_state) {
...
+    '#default_value' => $flood_config->get('flood_user_window'),
+  );
+  $form['user']['flood_uid_only'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Use only the User ID to identify users'),
+    '#description' => t('The default identifier is a combination of User ID and IP address. This is less secure but more resistant to denial-of-service attacks that could lock out all users with public user names. When enabled, checking for multiple login attempts will be based on the User ID only. This means that it will check regardless of the IP address from which the user is attempting to login.'),
+    '#default_value' => $flood_config->get('flood_uid_only'),
+  );
+
+  $form['actions']['#type'] = 'actions';
```
Comment (opi) [inline core/modules/user/user.admin.inc]: You can avoid this part, and the submit handler, by using https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.module/function/system_settings_form/1

**Comment by opi** [comment]
@herbdool here is a diff from a test, using `system_settings_form` and the `#config` key from FormAPI. It makes code cleaner, and avoid using a custom submit handler.

```diff
diff --git a/core/modules/user/user.admin.inc b/core/modules/user/user.admin.inc
index de9341ef8..897f8b80f 100755
--- a/core/modules/user/user.admin.inc
+++ b/core/modules/user/user.admin.inc
@@ -276,15 +276,19 @@ function user_admin_settings_submit($form, &$form_state) {
  *
  * @ingroup forms
  */
-function user_login_settings() {
+function user_login_settings($form, &$form_state) {
   $flood_config = config('user.flood');
   $core_config = config('system.core');
 
+  // Default config file
+  $form['#config'] = 'user.flood';
+
   // User login settings.
   $form['user_login'] = array(
     '#type' => 'fieldset',
     '#title' => t('User login'),
     '#weight' => -2,
+    '#config' => 'system.core', // config file override
   );
   $form['user_login']['user_login_method'] = array(
     '#type' => 'radios',
@@ -363,31 +367,7 @@ function user_login_settings() {
     '#default_value' => $flood_config->get('flood_uid_only'),
   );
 
-  $form['actions']['#type'] = 'actions';
-  $form['actions']['submit'] = array(
-    '#type' => 'submit',
-    '#value' => t('Save configuration'),
-  );
-
-  return $form;
-}
-
-/**
- * Submit handler for user_login_settings().
- */
-function user_login_settings_submit($form, &$form_state) {
-  config('user.flood')
-    ->set('flood_uid_only', $form_state['values']['flood_uid_only'])
-    ->set('flood_ip_limit', $form_state['values']['flood_ip_limit'])
-    ->set('flood_ip_window', $form_state['values']['flood_ip_window'])
-    ->set('flood_user_limit', $form_state['values']['flood_user_limit'])
-    ->set('flood_user_window', $form_state['values']['flood_user_window'])
-    ->save();
-  config('system.core')
-    ->set('user_login_method', $form_state['values']['user_login_method'])
-    ->save();
-
-  backdrop_set_message(t('The configuration options have been saved.'));
+  return system_settings_form($form);
 }
 
 /**
```

**Comment by herbdool** [comment]
@opi just testing it now. Going by the [docs](https://api.backdropcms.org/converting-modules-from-drupal), it seems that it wasn't possible to use that approach when saving to multiple config files.

> If you need to save values into multiple config files, we recommend that you add your own submit button, and submit handler for the form instead, as follows.

**Comment by opi** [comment]
hmm, docs probably needs to be patched, the doc of `system_settings_form` is more up to date https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.module/function/system_settings_form/1

The only drawback I saw was that a root `$form['#config']` is required, I was not able to only define a `#config` on each fieldset.

btw, thanks for implementing my patch <3

File: core/modules/user/user.admin.inc [hunk 2]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+      USER_LOGIN_USERNAME_OR_EMAIL => t('Username or email address'),
+      USER_LOGIN_USERNAME_ONLY => t('Username'),
+      USER_LOGIN_EMAIL_ONLY => t('Email address'),
+    ),
+    '#default_value' => $core_config->get('user_login_method'),
+  );
+
+  $form['ip_limit_settings'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Limit login attempts by IP address'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc:305]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 3]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+  );
+  $form['ip_limit_settings']['wrapper'] = array(
+    '#type' => 'container',
+    '#attributes' => array('class' => array('container-inline')),
+  );
+  $form['ip_limit_settings']['wrapper']['flood_ip_limit'] = array(
+    '#type' => 'select',
+    '#title' => t('Attempted login limit'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc:313]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 4]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+    '#title_display' => 'invisible',
+    '#options' => backdrop_map_assoc(array(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 75, 100, 125, 150, 200, 250, 500)),
+    '#default_value' => $flood_config->get('flood_ip_limit'),
+    '#prefix' => t('Limit to'),
+  );
+  $form['ip_limit_settings']['wrapper']['flood_ip_window'] = array(
+    '#type' => 'select',
+    '#title' => t('Attempted login time window'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc:321]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 5]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+    '#default_value' => $flood_config->get('flood_ip_window'),
+    '#prefix' => t('login attempts from one IP address per'),
+  );
+  $form['ip_limit_settings']['help'] = array(
+    '#type' => 'item',
+    '#description' => t('Do not allow any login from the current user\'s IP if the limit has been reached. This is independent of the per-user limit to catch attempts from one IP to log in to many different user accounts. By default we have a reasonably high limit since there may be only one apparent IP for all users at an institution.'),
+  );
+  $form['user_limit_settings'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Limit login attempts by a user'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 6]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+  );
+  $form['user_limit_settings']['wrapper'] = array(
+    '#type' => 'container',
+    '#attributes' => array('class' => array('container-inline')),
+  );
+  $form['user_limit_settings']['wrapper']['flood_user_limit'] = array(
+    '#type' => 'select',
+    '#title' => t('Attempted login limit'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc:341]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 7]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+    '#title_display' => 'invisible',
+    '#options' => backdrop_map_assoc(array(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 75, 100, 125, 150, 200, 250, 500)),
+    '#default_value' => $flood_config->get('flood_user_limit'),
+    '#prefix' => t('Limit to'),
+  );
+  $form['user_limit_settings']['wrapper']['flood_user_window'] = array(
+    '#type' => 'select',
+    '#title' => t('Attempted login time window'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc:349]: `login` => `log in`

File: core/modules/user/user.admin.inc [hunk 8]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+    '#title_display' => 'invisible',
+    '#options' => array(0 => t('None (disabled)')) + backdrop_map_assoc(array(60, 180, 300, 600, 900, 1800, 2700, 3600, 10800, 21600, 32400, 43200, 86400), 'format_interval'),
+    '#default_value' => $flood_config->get('flood_user_window'),
+    '#prefix' => t('login attempts by a user per'),
+  );
+  $form['user_limit_settings']['help'] = array(
+    '#type' => 'item',
+    '#description' => t('Configure the limit and the time window for users attempting to log in. That is, how many failed login attempts are allowed every X hours.'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc]: `login attempts` => `log in attempts`

File: core/modules/user/user.admin.inc [hunk 9]
```diff
@@ -288,6 +271,110 @@ function user_admin_settings_submit($form, &$form_state) {
...
+  );
+  $form['user_limit_settings']['flood_uid_only'] = array(
+    '#type' => 'radios',
+    '#title' => t('Identify users attempting to log in, using'),
+    '#options' => array(
+      FALSE => t('User ID and IP address combination'),
+      TRUE => t('User ID only'),
```
Comment (quicksketch) [inline core/modules/user/user.admin.inc]: I don't think these booleans end up saving as booleans in the config file -- a limitation of `system_settings_form()`, so I'd recommend this be made a bit less confusing to read by using options `'0'` and `'1'` (as strings) so they read like other `#options`.

File: core/modules/user/user.module [hunk 10]
```diff
@@ -1268,6 +1268,16 @@ function user_menu() {
     'file' => 'user.admin.inc',
   );
 
+  // Login settings.
+  $items['admin/config/people/login'] = array(
+    'title' => 'Login settings',
+    'description' => 'Configure login settings',
```
Comment (quicksketch) [inline core/modules/user/user.module]: I think these `login` uses are fine. As the description can be displayed in admin pages, this should be a sentence like `Configure the way visitors may log in and limit log in attempts.`

**Review by quicksketch — DISMISSED** [review]
Technically, looks great! I'm sorry I'm nitpicking terminology here. I don't feel strongly about this `login` vs. `log in`, but in some places we're not even consistent within the same sentence.

File: core/modules/user/user.module (hunk 10 above)
Comment (klonos) [reply core/modules/user/user.module]: Agree 👍 ...perhaps the second one should be a noun "login":

>Configure the way visitors may log in, and limit login attempts.

**Comment by herbdool** [comment]
@quicksketch made two small changes from review.

**Comment by quicksketch** [comment]
Looks great! Merged into 1.x for 1.11.0.

**Pull request merged**